api/v1/storage_autoscaler.go

```diff
@@ -0,0 +1,137 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// StorageAutoScalerSpec defines the desired state of StorageAutoScaler
+type StorageAutoScalerSpec struct {
+	// StorageCluster is the name of the storage cluster for which the storage scaling is to be done.
+	// +kubebuilder:validation:Required
+	StorageCluster corev1.LocalObjectReference `json:"storageCluster,omitempty"`
+	// DeviceClass is the name of the device class for which the storage scaling is to be done.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default="ssd"
+	DeviceClass string `json:"deviceClass,omitempty"`
+	// StorageCapacityLimit is the total aggregate capacity limit for the storage scaling for the specific deviceClass and storagecluster.
+	// +kubebuilder:validation:Required
+	StorageCapacityLimit resource.Quantity `json:"storageCapacityLimit,omitempty"`
+	// MaxOsdSize is the maximum size that Osd disk can be expanded to.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default="8Ti"
+	MaxOsdSize resource.Quantity `json:"maxOsdSize,omitempty"`
+	// StorageScalingThresholdPercent is the threshold percentage of the storage capacity that triggers the auto-scaling of the OSDs.
+	// Should be less than the OsdNearFullThresholdPercentage.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=70
+	StorageScalingThresholdPercent int `json:"storageScalingThresholdPercent,omitempty"`
+	// TimeoutSeconds is the time in seconds after which the storage auto-scaler will alert the user that the scaling operation has been failed.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=1800
+	TimeoutSeconds int `json:"timeoutSeconds,omitempty"`
+}
+
+type StorageAutoScalerPhase string
+
+const (
+	StorageAutoScalerPhaseNotStarted StorageAutoScalerPhase = "NotStarted"
+	StorageAutoScalerPhaseInProgress StorageAutoScalerPhase = "InProgress"
+	StorageAutoScalerPhaseSucceeded  StorageAutoScalerPhase = "Succeeded"
+	StorageAutoScalerPhaseFailed     StorageAutoScalerPhase = "Failed"
+)
+
+// TimestampedError is the error message with the time stamp.
+type TimestampedError struct {
+	// Message is the error message in case the storage scaling operation has failed.
+	Message string `json:"message,omitempty"`
+	// Timestamp is the time stamp when the error occurred.
+	Timestamp metav1.Time `json:"timestamp,omitempty"`
+}
+
+// StorageAutoScalerStatus defines the observed state of StorageAutoScaler
+type StorageAutoScalerStatus struct {
+	// Phase describes the Phase of StorageAutoScaler
+	// +kubebuilder:default="NotStarted"
```

Comment:
I'm not sure about a default here. Personally, I think `""` is a useful status for indicating when the controller hasn't acknowledged the resource yet. This is probably not huring anything, but it does subtly change an shared design pattern.

Reply:
As it is against the pattern and it should be ack by the controller, removed the default,

Made the status.spec optional, hope it is fine?